### ISSUE TITLE
libzim is not able to deal properly with an empty ZIM file

### ISSUE THREAD

**Issue by julianharty** [openzim/libzim#106]
As requested by @mgautierfr I'm raising this issue here. 
The original is https://github.com/kiwix/kiwix-android/issues/326 
---------
I believe we're already aware of the behaviour, here's an example that triggered the application dying when it was trying to open a zim file.

I've got a 64GB SD-card formatted with FAT32 (so the older tablet can read the content) with 2 large ZIM files: a fairly old Spanish wikipedia that's been on the card and in the device for a long time, and a newer ZIM file - the german novid wikipedia content I've just added directly (not downloaded using Kiwix Android).

The device is a Lenovo Yoga Tab B6000-H running Android 4.2.2 

I'll try testing on other devices that can accept the SD-Card.

Here's an extract of adb logcat. The jdwp messages are flooding log BTW
```
I/kiwix: Opening ZIM file /storage/sdcard1/wikipedia_es_all_05_201.zimaa
A/libc: ../../SOURCE/libzim/src/file_reader.cpp:90: virtual void zim::FileReader::read(char*, zim::offset_type, zim::offset_type) const: assertion "offset < _size" failed
D/jdwp: sendRequest : Len=0x56
A/libc: Fatal signal 11 (SIGSEGV) at 0xdeadbaad (code=1), thread 18322 (Thread-863)
```

The same Spanish ZIM file can be read by an older version of Kiwix Android (version 1.99) installed in a Nook HD+ running CyanogenMod 6.0.1

Reproduced this crash when I created a zero-byte `test.zim` on a Barnes & Noble Nook HD+ (Android 6.0.1). I doubt the specific device is that relevant though. Anyway, the same assertion failed.
The file was created using a file manager. `touch test.zim` would also create a similar file (run from `adb shell` etc.)

```
12-15 12:15:48.278 22482-22549/org.kiwix.kiwixmobile D/kiwix: Retrieving: content://org.kiwix.kiwixmobile.zim.base/-/j/body.js
12-15 12:15:48.283 22482-22528/org.kiwix.kiwixmobile W/ContentResolver: Failed to get type for: content://org.kiwix.kiwixmobile.zim.base/-/j/body.js (Attempt to invoke virtual method 'boolean java.lang.String.isEmpty()' on a null object reference)
12-15 12:15:48.283 22482-22528/org.kiwix.kiwixmobile W/ContentResolver: Failed to get type for: content://org.kiwix.kiwixmobile.zim.base/-/j/body.js (Attempt to invoke virtual method 'boolean java.lang.String.isEmpty()' on a null object reference)
12-15 12:15:48.283 22482-22554/org.kiwix.kiwixmobile D/kiwix: reading  -/j/body.js(mime: application/javascript, size: 17200) finished.
12-15 12:15:49.271 22482-22482/org.kiwix.kiwixmobile D/kiwix: Loaded URL: content://org.kiwix.kiwixmobile.zim.base/A/index.html
12-15 12:15:49.297 22482-22482/org.kiwix.kiwixmobile D/kiwix: Loaded URL: content://org.kiwix.kiwixmobile.zim.base/A/index.html
12-15 12:15:50.145 22482-22482/org.kiwix.kiwixmobile I/ListPopupWindow: Could not find method setEpicenterBounds(Rect) on PopupWindow. Oh well.
12-15 12:15:52.890 22482-22482/org.kiwix.kiwixmobile D/kiwix: onPause Save currentzimfile to preferences: /storage/6665-3832/Android/data/org.kiwix.kiwixmobile/wikipedia_en_uganda_09_2014 - copy.zim
12-15 12:15:53.089 22482-22482/org.kiwix.kiwixmobile I/kiwix: ZimManageActivity successfully bootstrapped
12-15 12:15:53.263 22482-22558/org.kiwix.kiwixmobile A/libc: ../../SOURCE/libzim/src/file_reader.cpp:90: virtual void zim::FileReader::read(char*, zim::offset_type, zim::offset_type) const: assertion "offset < _size" failed
12-15 12:15:53.308 22482-22557/org.kiwix.kiwixmobile D/android-storage-devices: Unable to create marker file, duplicates may be listed
12-15 12:15:53.431 22482-22557/org.kiwix.kiwixmobile D/android-storage-devices: Unable to create marker file, duplicates may be listed
12-15 12:15:53.443 22482-22558/org.kiwix.kiwixmobile W/google-breakpad: ### ### ### ### ### ### ### ### ### ### ### ### ###
12-15 12:15:53.443 22482-22558/org.kiwix.kiwixmobile W/google-breakpad: Chrome build fingerprint:
12-15 12:15:53.443 22482-22558/org.kiwix.kiwixmobile W/google-breakpad: 2.3
12-15 12:15:53.443 22482-22558/org.kiwix.kiwixmobile W/google-breakpad: 55
12-15 12:15:53.443 22482-22558/org.kiwix.kiwixmobile W/google-breakpad: 023f32a1-4862-425f-a49b-e5b1792614ed
12-15 12:15:53.443 22482-22558/org.kiwix.kiwixmobile W/google-breakpad: ### ### ### ### ### ### ### ### ### ### ### ### ###
12-15 12:15:53.443 22482-22558/org.kiwix.kiwixmobile E/chromium: ### WebView Version 44.0.2403.119 (code 246011900)
12-15 12:15:53.444 22482-22558/org.kiwix.kiwixmobile A/libc: Fatal signal 6 (SIGABRT), code -6 in tid 22558 (Thread-519)
12-15 12:15:53.461 22482-22557/org.kiwix.kiwixmobile D/kiwix: Searching directory /mnt
```

**Comment by kelson42** [comment]
@julianharty The naming of the file is quite strange `/storage/sdcard1/wikipedia_es_all_05_201.zimaa` it should be `/storage/sdcard1/wikipedia_es_all_05_2017.zimaa`

**Comment by kelson42** [comment]
The problem is reproducable with an empty file with:
```
$ zimdump -l test.zim 
error 2 opening file "test.zim
$ echo $?
254
```

**Comment by kelson42** [comment]
This ticket seems to be a duplicate of #48

**Comment by julianharty** [comment]
@kelson42  Perhaps the same fix would work for both cases. As a reminder, as a team, we'll need to make sure the Android app handles the error returned by the library elegantly.

**Comment by kelson42** [comment]
@mgautierfr It looks like this bug it at the source of many problem we are chasing for months on Android. I suspect that if it might be complicated to handle all kind of cases of corrupted data, it could be quite easy to handle empty files or files which doe not have the correct size (partially downloaded). Would that be possible to implement a small check before interpreting content:
1 - Check that the file is large enough to include the ZIM header (so at least 80 bytes?)
2 - Read the checksumPos, and check the overall file size is OK (should work with chunked ZIM files of course)?

**Comment by mgautierfr** [comment]
Fixed with PR #114 who add tests and lots of fix to handle this kind of problem.
